# Python for Data Science Practicce Session 1: Social Sciences

## How is Coronavirus Affecting Lives Across the World?

The [COVIDiSTRESS global survey](https://osf.io/z39us/) is an international collaborative undertaking for gathering data on human experiences, behaviour and attitudes during the COVID-19 pandemic. The survey focuses on psychological stress, compliance with behavioural guidelines to slow the spread of Coronavirus, and trust in governmental institutions and their preventive measures. 

In this notebook we are going to use the results of the COVIDiSTRESS survey conducted in April–May 2020 to investigate the impact of the Coronavirus pandemic on the perceived level of stress by individuals across the world as well as explore other related factors. 

Before we get started we need to import the `pandas` packages

In [87]:
# Import pandas
import pandas as pd

## Data Preprocessing

### Import

You can download the relevant dataset as a CSV file from this session's [materials](https://education.wdss.io/python-for-data-science/session-one/).

This dataset is more complex than the one we looked at in the teaching session and so we will need to acquaint ourselves with some more optional arguments of the `read_csv` function. Namely, look at the [documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) of the function, to see what the following parameters are for:

- `low_memory`
- `parse_dates`
- `dayfirst`
- `index_col`

Use this knowledge to import the dataset and assign it to the variable `raw_df`. Pay close attention to the following points:

- The dataset has a date column `RecordedDate` which uses day first format
- Because many columns have lots of missing values, there will likely be mixed type inference (a bad thing) when using low memory mode
- The first column of the dataset consists of row numbers which should be read as an index

In [88]:
raw_df = pd.read_csv('data/covidstress.csv',low_memory = False,parse_dates=[1],dayfirst = True,index_col=0)

raw_df

Duration..in.seconds.      RecordedDate UserLanguage  Dem_age  \
1                       180  30/05/2020 23:47          SAR       29   
2                      3100  29/05/2020 23:30           UR       20   
3                       127  30/05/2020 22:40          SAR       47   
4                      1710  29/05/2020 22:47           BG       79   
5                      2239  29/05/2020 22:42          SAR       61   
...                     ...               ...          ...      ...   
65531                   726  03/04/2020 14:35           FI       43   
65532                   918  04/04/2020 14:38           FI       31   
65533                  2662  03/04/2020 15:07           FI       69   
65534                  1309  03/04/2020 14:45          SAR       36   
65535                   957  03/04/2020 14:39           FI       46   

      Dem_gender                                            Dem_edu  \
1         Female                   College degree, bachelor, master   
2           Male                   College degree, bachelor, master   
3         Female  Some College, short continuing education or eq...   
4           Male                   College degree, bachelor, master   
5         Female  Some College, short continuing education or eq...   
...          ...                                                ...   
65531     Female                   College degree, bachelor, master   
65532     Female                   College degree, bachelor, master   
65533     Female                   College degree, bachelor, master   
65534     Female  Some College, short continuing education or eq...   
65535       Male                            Up to 6 years of school   

                      Dem_edu_mom      Dem_employment    Country Dem_Expat  \
1      Some College or equivalent        Not employed  Argentina       yes   
2                            None             Student   Pakistan       yes   
3      Some College or equivalent       Self-employed  Argentina        no   
4                  College degree        Not employed   Bulgaria        no   
5        Up to 12 years of school             Retired  Argentina        no   
...                           ...                 ...        ...       ...   
65531              College degree  Full time employed    Finland        no   
65532    Up to 12 years of school  Full time employed    Finland        no   
65533     Up to 9 years of school             Retired    Finland        no   
65534  Some College or equivalent  Full time employed  Argentina        no   
65535                         NaN        Not employed    Finland        no   

       ...                                         Final_open PSS10_avg  \
1      ...                                                NaN  2.900000   
2      ...                                                NaN  2.200000   
3      ...                                                NaN       NaN   
4      ...                                                NaN  3.600000   
5      ...                        Informaciones no confiables  2.714286   
...    ...                                                ...       ...   
65531  ...  Lemmikkini menehtyminen vuosi sitten lis‰‰ yks...  2.800000   
65532  ...                                                NaN  2.800000   
65533  ...                                                NaN  2.100000   
65534  ...                                                NaN  2.000000   
65535  ...                            Oma terveys on pett‰nyt  3.500000   

       SLON3_avg       neu       ext       ope       agr       con SPS_avg  \
1       3.000000       NaN       NaN       NaN       NaN       NaN     NaN   
2       2.333333  2.000000  5.000000  5.333333  5.000000  5.000000     5.0   
3            NaN       NaN       NaN       NaN       NaN       NaN     NaN   
4       4.000000  4.000000  4.666667  4.000000  5.000000  5.000000     3.9   
5       1.000000  3.666667  4.666667  5.333333  5.000000  4.666667     5

> **Bonus**
>
> Suppose we didn't trust the index column of the original dataset. How could we use the default index provide by `pandas` instead?

Let's explore the dataset by doing the following:
- Print the first 3 rows of the dataset
- Print the dimensions of the dataset
- Print numerical summaries of the columns
- Print the data types of the columns

In [89]:
# First 3 rows
raw_df.head(3)

Duration..in.seconds.      RecordedDate UserLanguage  Dem_age Dem_gender  \
1                   180  30/05/2020 23:47          SAR       29     Female   
2                  3100  29/05/2020 23:30           UR       20       Male   
3                   127  30/05/2020 22:40          SAR       47     Female   

                                             Dem_edu  \
1                   College degree, bachelor, master   
2                   College degree, bachelor, master   
3  Some College, short continuing education or eq...   

                  Dem_edu_mom Dem_employment    Country Dem_Expat  ...  \
1  Some College or equivalent   Not employed  Argentina       yes  ...   
2                        None        Student   Pakistan       yes  ...   
3  Some College or equivalent  Self-employed  Argentina        no  ...   

  Final_open PSS10_avg  SLON3_avg  neu  ext       ope  agr  con SPS_avg  \
1        NaN       2.9   3.000000  NaN  NaN       NaN  NaN  NaN     NaN   
2        NaN       2.2   2.333333  2.0  5.0  5.333333  5.0  5.0     5.0   
3        NaN       NaN        NaN  NaN  NaN       NaN  NaN  NaN     NaN   

  Scale_UCLA_TRI_avg  
1                NaN  
2                NaN  
3                NaN  

[3 rows x 152 columns]

In [90]:
# Dimensions of the dataset
raw_df.shape

(65535, 152)

In [91]:
# Numerical summaries of the columns
raw_df.describe()

Dem_age  Dem_dependents  Dem_isolation_adults  Dem_isolation_kids  \
count  65535.000000    63317.000000          53379.000000        52463.000000   
mean      40.741375        0.933146              1.533937            0.484513   
std       14.322085        1.805861              3.231931            1.275804   
min       18.000000        0.000000              0.000000            0.000000   
25%       29.000000        0.000000              1.000000            0.000000   
50%       40.000000        0.000000              1.000000            0.000000   
75%       51.000000        2.000000              2.000000            1.000000   
max      110.000000      110.000000            110.000000          110.000000   

       Scale_PSS10_UCLA_1  Scale_PSS10_UCLA_2  Scale_PSS10_UCLA_3  \
count        60336.000000        60312.000000        60344.000000   
mean             2.556782            2.663450            3.078019   
std              1.094821            1.104996            1.112253   
min              1.000000            1.000000            1.000000   
25%              2.000000            2.000000            2.000000   
50%              3.000000            3.000000            3.000000   
75%              3.000000            3.000000            4.000000   
max              5.000000            5.000000            5.000000   

       Scale_PSS10_UCLA_4  Scale_PSS10_UCLA_5  Scale_PSS10_UCLA_6  ...  \
count        60274.000000        60253.000000        60347.000000  ...   
mean             3.683612            3.158266            2.683845  ...   
std              1.049348            0.968456            1.118127  ...   
min              1.000000            1.000000            1.000000  ...   
25%              3.000000            3.000000            2.000000  ...   
50%              4.000000            3.000000            3.000000  ...   
75%              4.000000            4.000000            3.000000  ...   
max              5.000000            5.000000            5.000000  ...   

       Expl_media_6     PSS10_avg     SLON3_avg           neu           ext  \
count   48753.00000  60811.000000  60673.000000  56884.000000  56858.000000   
mean        3.31034      2.628514      2.581236      3.316275      3.999156   
std         1.37873      0.735439      0.990043      1.050524      1.116921   
min         1.00000      1.000000      1.000000      1.000000      1.000000   
25%         2.00000      2.100000      2.000000      2.666667      3.333333   
50%         3.00000      2.600000      2.666667      3.333333      4.000000   
75%         4.00000      3.100000      3.333333      4.000000      5.000000   
max         6.00000      5.000000      5.000000      6.000000      6.000000   

                ope           agr           con       SPS_avg  \
count  56856.000000  56869.000000  56865.000000  49669.000000   
mean       4.478871      4.439443      4.380509      4.849583   
std        0.947898      0.832629      0.890780      0.907215   
min        1.000000      1.000000      1.000000      1.000000   
25%        4.000000      4.000000      3.666667      4.400000   
50%        4.666667      4.666667      4.333333      5.000000   
75%        5.000000      5.000000      5.000000      5.500000   
max        6.000000      6.000000      6.000000      6.000000   

       Scale_UCLA_TRI_avg  
count           51.000000  
mean             1.818627  
std              0.968060  
min              0.000000  
25%              1.000000  
50%              1.750000  
75%              2.500000  
max              4.000000  

[8 rows x 126 columns]

> **Bonus**
>
> When printing the above, we only get to see 10 columns in total. For printing out full datasets and series, checkout the [option context](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.option_context.html)

In [92]:
# Data types of columns
with pd.option_context("display.max_rows", raw_df.shape[1]):          #Here I have printed all of the column types 
    print(raw_df.dtypes)

Duration..in.seconds.     object
RecordedDate              object
UserLanguage              object
Dem_age                    int64
Dem_gender                object
Dem_edu                   object
Dem_edu_mom               object
Dem_employment            object
Country                   object
Dem_Expat                 object
Dem_state                 object
Dem_maritalstatus         object
Dem_dependents           float64
Dem_riskgroup             object
Dem_islolation            object
Dem_isolation_adults     float64
Dem_isolation_kids       float64
AD_gain                   object
AD_loss                   object
AD_check                  object
Scale_PSS10_UCLA_1       float64
Scale_PSS10_UCLA_2       float64
Scale_PSS10_UCLA_3       float64
Scale_PSS10_UCLA_4       float64
Scale_PSS10_UCLA_5       float64
Scale_PSS10_UCLA_6       float64
Scale_PSS10_UCLA_7       float64
Scale_PSS10_UCLA_8       float64
Scale_PSS10_UCLA_9       float64
Scale_PSS10_UCLA_10      float64
Scale_SLON

### Cleaning

There are 152 columns in the dataset, many of which we won't need in our analysis. Instead, we want to focus on the following variables.

| description | variable name | measurement |
| --- | --- | --- |
| date of record | `RecordedDate` | date "DD/MM/YYYY hh:mm"
| age | `Dem_age` | age in years
| gender | `Dem_gender` | Male/Female
| country | `Country` | country name
| employment status| `Dem_employment` | type of employment
| perceived stress for the past week | `Scale_PSS10_UCLA_1` to `Scale_PSS10_UCLA_10` | 1=never, 5=very often; 10 items
| trust in institutions | `OECD_insititutions_1` (government), `OECD_insititutions_4` (health system) | 0=not at all, 10=completely
| trust in country's preventive measures | `Trust_countrymeasure` | 0=too little, 5= appropriate, 11=too much

Stress was meassured in 10 different categories. Create a list, `stress_columns` of column names `Scale_PSS10_UCLA_X`, where `X` ranges from 1 to 10.

In [93]:
stress_columns = ['Scale_PSS10_UCLA_'+ str(x) for x in range(1,11)]
stress_columns

['Scale_PSS10_UCLA_1',
 'Scale_PSS10_UCLA_2',
 'Scale_PSS10_UCLA_3',
 'Scale_PSS10_UCLA_4',
 'Scale_PSS10_UCLA_5',
 'Scale_PSS10_UCLA_6',
 'Scale_PSS10_UCLA_7',
 'Scale_PSS10_UCLA_8',
 'Scale_PSS10_UCLA_9',
 'Scale_PSS10_UCLA_10']

We can then combine these with the other column names.

In [94]:
columns = [
    'RecordedDate', 'Dem_age', 'Dem_gender', 'Country', 'Dem_employment',
    'OECD_insititutions_1', 'OECD_insititutions_4', 'Trust_countrymeasure'
] + stress_columns

columns

['RecordedDate',
 'Dem_age',
 'Dem_gender',
 'Country',
 'Dem_employment',
 'OECD_insititutions_1',
 'OECD_insititutions_4',
 'Trust_countrymeasure',
 'Scale_PSS10_UCLA_1',
 'Scale_PSS10_UCLA_2',
 'Scale_PSS10_UCLA_3',
 'Scale_PSS10_UCLA_4',
 'Scale_PSS10_UCLA_5',
 'Scale_PSS10_UCLA_6',
 'Scale_PSS10_UCLA_7',
 'Scale_PSS10_UCLA_8',
 'Scale_PSS10_UCLA_9',
 'Scale_PSS10_UCLA_10']

Now, create a subset of this dataset, `covid_stress`, containing only columns of interest.

In [95]:
covid_stress = raw_df.loc[:,columns]
covid_stress

RecordedDate  Dem_age Dem_gender    Country      Dem_employment  \
1      30/05/2020 23:47       29     Female  Argentina        Not employed   
2      29/05/2020 23:30       20       Male   Pakistan             Student   
3      30/05/2020 22:40       47     Female  Argentina       Self-employed   
4      29/05/2020 22:47       79       Male   Bulgaria        Not employed   
5      29/05/2020 22:42       61     Female  Argentina             Retired   
...                 ...      ...        ...        ...                 ...   
65531  03/04/2020 14:35       43     Female    Finland  Full time employed   
65532  04/04/2020 14:38       31     Female    Finland  Full time employed   
65533  03/04/2020 15:07       69     Female    Finland             Retired   
65534  03/04/2020 14:45       36     Female  Argentina  Full time employed   
65535  03/04/2020 14:39       46       Male    Finland        Not employed   

       OECD_insititutions_1  OECD_insititutions_4  Trust_countrymeasure  \
1                       NaN                   NaN                   NaN   
2                       5.0                   2.0                   5.0   
3                       NaN                   NaN                   NaN   
4                       0.0                   4.0                   5.0   
5                       9.0                   9.0                   5.0   
...                     ...                   ...                   ...   
65531                   9.0                   9.0                   5.0   
65532                   7.0                   8.0                   5.0   
65533                   8.0                   9.0                   5.0   
65534                   5.0                   7.0                   2.0   
65535                   8.0                  10.0                   5.0   

       Scale_PSS10_UCLA_1  Scale_PSS10_UCLA_2  Scale_PSS10_UCLA_3  \
1                     5.0                 2.0                 5.0   
2                     3.0                 3.0                 1.0   
3                     NaN                 NaN                 NaN   
4                     3.0                 5.0                 3.0   
5                     3.0                 3.0                 2.0   
...                   ...                 ...                 ...   
65531                 2.0                 2.0                 4.0   
65532                 3.0                 3.0                 3.0   
65533                 2.0                 2.0                 3.0   
65534                 2.0                 3.0                 3.0   
65535                 3.0                 5.0                 4.0   

       Scale_PSS10_UCLA_4  Scale_PSS10_UCLA_5  Scale_PSS10_UCLA_6  \
1                     4.0                 4.0                 1.0   
2                     5.0                 2.0                 3.0   
3                     NaN                 NaN                 NaN   
4                     3.0                 3.0                 5.0   
5                     NaN                 3.0                 NaN   
...                   ...                 ...                 ...   
65531                 3.0                 2.0                 2.0   
65532                 4.0                 4.0                 3.0   
65533                 4.0                 4.0                 2.0   
65534                 5.0                 2.0                 1.0   
65535                 3.0                 3.0                 4.0   

       Scale_PSS10_UCLA_7  Scale_PSS10_UCLA_8  Scale_PSS10_UCLA_9  \
1                     3.0                 5.0                 3.0   
2                     4.0                 4.0                 2.0   
3                     NaN                 NaN                 NaN   
4                     2.0                 3.0                 3.0   
5                     5.0                 2.0                 NaN   
...                   ...                 ...                 ...   
65531                 3.0       

The original dataset is fairly large so we may want to remove it from memory now that we no longer need it.

In [96]:
# Clear up memory
del raw_df

Many fields in the dataset have missing/unknown values. We can use `covid_stress.isna()` to return a Boolean dataset where each value is `True` if the original value in that field was missing. Let's investigate where these missing values are.

> Remember, `True`/`False` are converted to `1`/`0` respectively when cast to integers. This means we can use the `sum` and `mean` methods on a Boolean dataset or series to count or find proportions of `True`'s

In [97]:
# How many missing values are there in each column?

[sum(covid_stress.iloc[:,x].isna()) for x in range(0, covid_stress.shape[1])]


[0,
 0,
 154,
 345,
 659,
 8074,
 7547,
 7304,
 5199,
 5223,
 5191,
 5261,
 5282,
 5188,
 5277,
 5298,
 5163,
 5190]

In [98]:
covid_stress.shape[1]


18

In [99]:
# What proportion of records (rows) have at least one missing value?
miss_per_row = [sum(covid_stress.iloc[x,:].isna()) for x in range(0, covid_stress.shape[0])]
sum(i >= 1 for i in miss_per_row)/covid_stress.shape[0]


0.1706111238269627

Given that we have 65535 records in total we can drop all the records with at least one missing value and still have many tens of thousands left. We will learn how to deal with missing values properly in later sessions.

To do this, we use the `dropna` method of the dataframe. Search for "pandas drop missing values" on the web to find the documentation for this. Save the cleaned dataset as `covid_stress_clean`

In [100]:
covid_stress_clean = covid_stress.dropna(how = 'any')  # This drop any rows that have any missing values

> **Bonus**
>
> Recall from the teaching session that many pandas operations can be done _inplace_. How would we drop rows with missing values in this fashion (check the documentation again if needed).

Let's take a random sample of 10 rows and verify that they have no missing values.

In [101]:
# Random sample of cleaned dataset
covid_stress_clean.sample(10)

RecordedDate  Dem_age Dem_gender      Country      Dem_employment  \
43037  06/04/2020 16:21       42     Female       Kosovo  Part time employed   
33244  08/04/2020 12:44       49     Female      Finland  Full time employed   
55522  04/04/2020 23:43       52       Male      Finland  Full time employed   
34548  08/04/2020 06:16       23     Female      Finland        Not employed   
22273  13/04/2020 12:24       57     Female       Sweden       Self-employed   
31087  09/04/2020 05:58       45       Male       Sweden  Full time employed   
24595  12/04/2020 12:36       49     Female       Sweden  Full time employed   
12981  19/04/2020 05:26       27     Female  Switzerland  Part time employed   
4454   12/05/2020 01:53       36       Male        Spain  Full time employed   
16424  18/04/2020 02:44       49     Female  Netherlands       Self-employed   

       OECD_insititutions_1  OECD_insititutions_4  Trust_countrymeasure  \
43037                   8.0                   7.0                  10.0   
33244                   7.0                   6.0                   5.0   
55522                   6.0                   9.0                   5.0   
34548                   6.0                   9.0                   4.0   
22273                  10.0                  10.0                   9.0   
31087                   8.0                   7.0                   5.0   
24595                   9.0                   9.0                   5.0   
12981                   7.0                   6.0                   5.0   
4454                    7.0                   6.0                   3.0   
16424                   8.0                   9.0                   7.0   

       Scale_PSS10_UCLA_1  Scale_PSS10_UCLA_2  Scale_PSS10_UCLA_3  \
43037                 3.0                 4.0                 3.0   
33244                 2.0                 2.0                 2.0   
55522                 1.0                 1.0                 2.0   
34548                 1.0                 4.0                 3.0   
22273                 2.0                 3.0                 3.0   
31087                 1.0                 5.0                 2.0   
24595                 1.0                 5.0                 2.0   
12981                 3.0                 3.0                 3.0   
4454                  2.0                 2.0                 2.0   
16424                 1.0                 2.0                 2.0   

       Scale_PSS10_UCLA_4  Scale_PSS10_UCLA_5  Scale_PSS10_UCLA_6  \
43037                 5.0                 3.0                 4.0   
33244                 5.0                 4.0                 2.0   
55522                 5.0                 5.0                 1.0   
34548                 3.0                 3.0                 3.0   
22273                 5.0                 2.0                 1.0   
31087                 2.0                 3.0                 3.0   
24595                 5.0                 4.0                 2.0   
12981                 4.0                 4.0                 2.0   
4454                  5.0                 2.0                 2.0   
16424                 4.0                 4.0                 1.0   

       Scale_PSS10_UCLA_7  Scale_PSS10_UCLA_8  Scale_PSS10_UCLA_9  \
43037                 4.0                 3.0                 3.0   
33244                 4.0                 5.0                 1.0   
55522                 4.0                 5.0                 1.0   
34548                 4.0                 4.0                 2.0   
22273                 4.0                 2.0                 3.0   
31087                 5.0                 4.0                 2.0   
24595                 5.0                 5.0                 1.0   
12981                 4.0                 4.0                 3.0   
4454                  5.0                 4.0                 1.0   
16424                 4.0                 4.0                 1.0   

       Scale_PSS10_UCLA_

> Note, the `covid_stress` dataset isn't too large so we don't need to worry about deleting it. Sometimes it's worth holding onto these things in case we need to go back to them and don't want to rerun cells. How big is the dataset exactly? Have a look at the documentation for the `info` method, which can answer this.

In [102]:
covid_stress.info()   # Output gives the memory usage at the end

<class 'pandas.core.frame.DataFrame'>
Int64Index: 65535 entries, 1 to 65535
Data columns (total 18 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   RecordedDate          65535 non-null  object 
 1   Dem_age               65535 non-null  int64  
 2   Dem_gender            65381 non-null  object 
 3   Country               65190 non-null  object 
 4   Dem_employment        64876 non-null  object 
 5   OECD_insititutions_1  57461 non-null  float64
 6   OECD_insititutions_4  57988 non-null  float64
 7   Trust_countrymeasure  58231 non-null  float64
 8   Scale_PSS10_UCLA_1    60336 non-null  float64
 9   Scale_PSS10_UCLA_2    60312 non-null  float64
 10  Scale_PSS10_UCLA_3    60344 non-null  float64
 11  Scale_PSS10_UCLA_4    60274 non-null  float64
 12  Scale_PSS10_UCLA_5    60253 non-null  float64
 13  Scale_PSS10_UCLA_6    60347 non-null  float64
 14  Scale_PSS10_UCLA_7    60258 non-null  float64
 15  Scale_PSS10_UCLA_8 

### Feature Engineering

Before we continue, we will make a copy of the current dataframe. This will make it easier to return to this point by rerunning this cell if any mistakes are made.

In [103]:
covid_df = covid_stress_clean.copy()

The levels of stress were measured in 10 different categories on a scale from 1=never, 5=very often. We want analyse only aggregated information about the stress level, i.e. the mean of those 10 values. For this, we create a new column `Avg_stress`

In [104]:
covid_df.columns

Index(['RecordedDate', 'Dem_age', 'Dem_gender', 'Country', 'Dem_employment',
       'OECD_insititutions_1', 'OECD_insititutions_4', 'Trust_countrymeasure',
       'Scale_PSS10_UCLA_1', 'Scale_PSS10_UCLA_2', 'Scale_PSS10_UCLA_3',
       'Scale_PSS10_UCLA_4', 'Scale_PSS10_UCLA_5', 'Scale_PSS10_UCLA_6',
       'Scale_PSS10_UCLA_7', 'Scale_PSS10_UCLA_8', 'Scale_PSS10_UCLA_9',
       'Scale_PSS10_UCLA_10'],
      dtype='object')

In [105]:
# Create new column by averaging all stress columns
covid_df['Avg_stress'] = covid_df[stress_columns].mean(axis = 1)


# Drop the now redundant stress columns
covid_df = covid_df.drop(stress_columns, axis = 1)

The column names `OECD_insititutions_1`, `OECD_insititutions_4` are not the most friendly. Let's rename them to something more intuitive.

In [106]:
mapper = {
    "OECD_insititutions_1" : "Trust_gov",
    "OECD_insititutions_4" : "Trust_health"
}
covid_df = covid_df.rename(mapper, axis=1)
covid_df

RecordedDate  Dem_age                  Dem_gender    Country  \
2      29/05/2020 23:30       20                        Male   Pakistan   
4      29/05/2020 22:47       79                        Male   Bulgaria   
6      29/05/2020 21:25       68                        Male      Italy   
7      29/05/2020 21:25       29  Other/would rather not say  Argentina   
8      29/05/2020 21:25       38                      Female  Argentina   
...                 ...      ...                         ...        ...   
65531  03/04/2020 14:35       43                      Female    Finland   
65532  04/04/2020 14:38       31                      Female    Finland   
65533  03/04/2020 15:07       69                      Female    Finland   
65534  03/04/2020 14:45       36                      Female  Argentina   
65535  03/04/2020 14:39       46                        Male    Finland   

           Dem_employment  Trust_gov  Trust_health  Trust_countrymeasure  \
2                 Student        5.0           2.0                   5.0   
4            Not employed        0.0           4.0                   5.0   
6                 Retired        7.0           8.0                   5.0   
7      Part time employed        3.0           4.0                   5.0   
8            Not employed        3.0           3.0                   4.0   
...                   ...        ...           ...                   ...   
65531  Full time employed        9.0           9.0                   5.0   
65532  Full time employed        7.0           8.0                   5.0   
65533             Retired        8.0           9.0                   5.0   
65534  Full time employed        5.0           7.0                   2.0   
65535        Not employed        8.0          10.0                   5.0   

       Avg_stress  
2             2.8  
4             3.4  
6             2.9  
7             2.9  
8             3.4  
...           ...  
65531         2.6  
65532         3.2  
65533         2.9  
65534         2.8  
65535         3.5  

[54354 rows x 9 columns]

We're now ready to start answering questions about the dataset. Let's print out the data types of the columns to check that everything is as expected before we continue.

In [107]:
covid_df.dtypes

RecordedDate             object
Dem_age                   int64
Dem_gender               object
Country                  object
Dem_employment           object
Trust_gov               float64
Trust_health            float64
Trust_countrymeasure    float64
Avg_stress              float64
dtype: object

Check that the variable types of each of the column is as expected

## Data Analysis

### High-level Questions

When the study was conducted? (i.e. what is the range of dates in the records?)

In [109]:
# Minimum and maximum dates in dataset
covid_df[['RecordedDate']].max()


RecordedDate    30/05/2020 20:14
dtype: object

In [113]:
covid_df[['RecordedDate']].min()


RecordedDate    01/05/2020 00:33
dtype: object

What proportion of observations are from Finland, Sweden, or Norway? Use `.round()` to print the answer to 3 decimal places.

In [ ]:
# Proportion of Finish, Swedish, Norwegian respondents


Print the dataset sorted by age in descending order.

In [ ]:
# Sort by decreasing age


What about the median value for age? Print your answer as part of a meaningful sentence.

In [ ]:
# Median age of respondents
print("The median age of respondents is", covid_df.Dem_age.median())

### Characteristics of Respondents

What is the gender breakdown of respondents? Use the `value_counts` method to find out (after searching for its documentation of course).

In [ ]:
# Gender breakdown


It would be nicer to see percentages rather than absolute counts. There are many ways to do this. Take your pick or write down as many as you can think of.

In [ ]:
# Gender breakdown as proportions


Suppose we want to divide the respondents in 4 age of equal width, how many respondents would land in each group? (This can also be performed using `value_counts` too)

In [ ]:
# Binned value counts


What proportion of the observations have each employment type, excluding students and retirees?

In [ ]:
# Employment types for non-students/retirees


### Stress During the COVID-19 Pandemic

We can perform grouped summaries using the `groupby` method, documented [here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html). These returned a grouped dataframe, which we can apply aggregation methods such as `mean` and `min` to, in which case the aggregation is applied separately to each group. An example is given below.

In [ ]:
raw_df.groupby('Dem_edu')

In [ ]:
# First observation for each gender
covid_df.groupby('Dem_gender')['RecordedDate'].min()

Identify the 10 countries with the highest average value of stress.

How much can we trust these values? Are there enough respondents from each of those countries to make an an inference about the level of stress of the entire country? It would be helpful to aggregate both the mean stress level and observation count. We can do this using the `agg` method. Read the documentation for this then use it to recreate the above result with counts.

Suprising? Now let's identify 10 countries with highest average value of stress with at least 200 respondents.

In [ ]:
country_df[country_df['Count'] >= 200] \
    .sort_values('Mean_stress', ascending=False).head(10)

To extend this comparison, we wish to create a dateset `country_fltr` which contains the mean stress, trust in health organisations, government and health measures, and the observation count for all countries with at least 200 observations.

In [ ]:
country_df = covid_df.groupby('Country').agg(
    Mean_stress=('Avg_stress', 'mean'),
    Mean_gov_trust=('Trust_gov', 'mean'),
    Mean_health_trust=('Trust_health', 'mean'),
    Mean_measures_trust=('Trust_countrymeasure', 'mean'),
    Count=('Country', 'count')
)

country_fltr = country_df[country_df.Count>200]

Now sort `country_fltr` by different categories and in ascending/descending order, to explore these 4 measure across the countries.

It might be interesting to compare the trust in country's health system with the average stress level. One way to do it is to make a scatter plot with the mean trust in health on the x-axis and mean level of stress on the y-axis. (More about plotting in next session, but I couldn't resist making at least one plot in this notebook)

In [ ]:
country_fltr.plot(x='Mean_health_trust', y='Mean_stress', kind='scatter');

Cool! It seems like there is a negative correlation between these two variables. Can you spot the outlier? Find out which country this datapoint corresponds to. 

### A Closer Look at the UK

In this section we will draw comparisons between the United Kingdom and the rest of Europe. We have provided an Excel spreadsheet in the [session materials](https://education.wdss.io/python-for-data-science/session-one/) that maps countries to continents, which can be used to aid in this comparison. Start by loading this dataset into the notebook.

In [ ]:
# Load the country mapping spreadsheet


Create a pandas series of all European countries except the United Kingdom.

Use this series to create two dataframes, `uk_df` and `euro_df` containing observations from the UK and rest of Europe respectively.

In [ ]:
uk_df = # ...
euro_df = # ...

Is the average stress level in the UK above or below average of the rest of Europe? By how much? 

What proportion of people in the UK aged between 40 and 60 (inclusive) reported an average stress level above 3? This is a longer question so try to break it up into sensible chunks.

What the average stress level for each employment status and gender pair in Europe? (Note: you can use a list of labels when grouping a dataframe)

> Wow! What happened there? We've just been introduced to the pandas multi-index, a relatively unique feature of the package which offers incredible power and flexibility. This is considered an advanced feature and so we will not go into any more detail about it but you can read more [here](https://pandas.pydata.org/pandas-docs/stable/user_guide/advanced.html).

Did the trust in government tend to increase/decrease or remain stable over the period considered in the report? Compare the mean trust in government in first 3 weeks of April starting from 2020-04-06.

You can do this using base Python `for` loops. Loop through through consecutive pairs of days in the provided date list and calculate the mean trust in government for an subset of the dataset filtered to only have days between the current date in the list and 7 days in the future.

This isn't the most elegant and is certainly not the most efficient. A better solution is below, though it goes to show that pandas and base Python can be used together when needed.

In [ ]:
uk_df.groupby(uk_df.RecordedDate.dt.isocalendar().week).mean()['Trust_gov']

## Your Time to Shine

Come up with some new questions about this dataset and try to answer them using your newly obtained skills.